dlrover/python/elastic_agent/torch/ckpt_saver.py

```diff
@@ -393,6 +394,304 @@ def save_shm_to_storage(self):
+            d.close()
+
+        for lock in self._shm_lock:
+            lock.close()
+
+        self._executor.shutdown()
+
+    def _sync_shm_to_storage(self):
+        """
+        The loop to persist the state dict from the memory
+        buffer into the storage.
+        """
+        logger.info(
+            "ShardingSaver Start saving the checkpointing state dict to storage."
+        )
+
+        while True:
+            step = self._to_save_queue.get()
+            logger.info(
+                "ShardingSaver save checkpoint to storage, step: %s", step
+            )
+            self._save_shm_to_storage(step)
+
+    def _save_shm_to_storage(self, step):
+        logger.info(
+            f"Rank {self._node_rank} start save checkpoint to storage, step: {step}"
+        )
+        self._writing_storage = True
+        ckpt_path = self._get_ckpt_path(step)
+
+        if os.path.exists(ckpt_path):
+            logger.info(f"Checkpoint for step {step} already exists, skip")
+            self._writing_storage = False
+            return
+
+        def _save_stage(local_rank: int, write_path: str):
+            try:
+                if not self._tensor_shm[local_rank]:
+                    self._tensor_shm[local_rank] = SharedMemory(
+                        name=self._shm_name[local_rank]
+                    )
+
+                self._shm_lock[local_rank].acquire()
+                logger.info(
+                    f"Local rank {local_rank} Save checkpoint from the shared memory "
+                    f"into the storage {write_path}."
+                )
+                meta_dict = self._shared_ckpt_meta[local_rank].get()
+                state_dict = _read_state_dict_from_shm(
+                    meta_dict, self._tensor_shm[0]
+                )
+
+                self._persist_to_storage(state_dict, write_path, step)
+                self._shm_lock[local_rank].release()
+                return True
+
+            except Exception as e:
+                logger.error(
+                    f"Rank {local_rank} save checkpoint failed, error: {e}",
+                    exc_info=True,
+                )
+                self._shm_lock[local_rank].release()
+                return False
+
+        stage_path = os.path.join(self._get_stage_path(), str(step))
+        os.makedirs(stage_path, exist_ok=True)
+
+        step_done_path = os.path.join(
+            self._get_stage_path(), str(step) + ".done"
+        )
+        os.makedirs(step_done_path, exist_ok=True)
+
+        step_done_file = os.path.join(step_done_path, str(self._node_rank))
+
+        write_success = False
+        if os.path.exists(step_done_file):
+            logger.info(f"Rank {self._node_rank} already done for step {step}")
+            write_success = True
+        else:
+            # save to stage path for each local rank
+            futures = []
+            for i in range(self.num_proc):
+                future = self._executor.submit(_save_stage, i, stage_path)
```

Comment:
Is saving in parallel with threads faster than serial saving? If the bottleneck is the IO speed to write the storage, saving in parallel may not significantly improve the performance.